### DISCUSSION TITLE
Claude desktop app bug with MCPs: passing in object parameters for JSON schema defined with allOf

### DISCUSSION THREAD

**Discussion by domdomegg** [modelcontextprotocol/.github#90] [General]
### Pre-submission Checklist

- [X] I have verified that this discussion would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions to avoid duplicates

### Discussion Topic

tldr: In the Claude desktop app, when using tools with parameter schemas that include allOf combined with object types, the parameter is converted to a string instead of being passed as an object, even when properly formatted in the XML tags.

## Description

When using tools with parameter schemas that include `allOf` combined with object types, the parameter is being converted to a string instead of being passed as an object, even when properly formatted in the XML tags.

## Schema Example

The problematic tool input schema pattern looks like:

```json
{
  "parameters": {
    "properties": {
      "field": {
        "allOf": [
          {
            "properties": {
              "name": { "type": "string" },
              "description": { "type": "string" }
            },
            "required": ["name"],
            "type": "object"
          },
          {
            "anyOf": [
              {
                "properties": {
                  "type": { "const": "multipleRecordLinks" },
                  "options": {
                    "properties": {
                      "linkedTableId": { "type": "string" }
                    }
                  }
                },
                "required": ["type", "options"],
                "type": "object"
              }
              // ... other field types
            ]
          }
        ]
      }
    }
  }
}
```

## Reproduction Steps
1. Use a tool with an `allOf` schema containing object types (e.g. create_field)
2. Pass a properly formatted object:
```xml
<parameter name="field">{"name": "MVP", "type": "multipleRecordLinks", "options": {"linkedTableId": "tbljmSetpBaI6qGT6"}}</parameter>
```
3. Observe error: "Expected object, received string"

## Error Message

```
"Error in tool create_field: [
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": ["field"],
    "message": "Expected object, received string"
  }
]"
```

## Expected Behavior
Parameters matching an `allOf` schema with object types should be passed as objects, not converted to strings.

## System Information

- Claude Version: Claude 3.5 Sonnet (20241022)
- Tool: create_field with allOf schema

## Additional Notes

This appears specific to how the Anthropic app wrapper handles parameters when the schema uses `allOf` with object types. The parameter value is being stringified somewhere in the processing pipeline before reaching the tool.

**Comment by domdomegg** [comment]
This actually seems wider: any top-level allOf or anyOf with an object seems to have a problem. Here's an MRE: https://github.com/domdomegg/mcp-90-mre/blob/master/src/index.ts

And a video of Claude struggling with it:

https://github.com/user-attachments/assets/e22b0d06-f463-4c74-a081-cb0a631cb5e9

**Reply by domdomegg** [reply]
Chats as text:

---

I'll help you format those messages using the available formatting tools. I'll try them one by one.

First, let's try a hello greeting:

I apologize for the error. Let me correct the format and try again:

I apologize for the continued errors. Let me try a different approach:

I apologize for the continued errors. It seems I'm having trouble with the correct format for these function calls. Let me try one last time with the sorry message:

I was able to successfully format a sorry message! However, I apologize that I wasn't able to get the hello and goodbye messages working correctly. It seems there might be an issue with how I'm formatting the parameters for those functions. Would you like me to try again with a different format for the hello and goodbye messages?

---

I'll show you the exact XML I tried for each command, with HTML escaping:

For the first hello attempt:
```
&lt;antml:function_calls&gt;
&lt;antml:invoke name="format_hello"&gt;
&lt;antml:parameter name="obj">{"type": "individual", "person": "friend", "timeOfDay": "morning"}&lt;/antml:parameter&gt;
&lt;/antml:invoke&gt;
&lt;/antml:function_calls&gt;
```

For the second hello attempt:
```
&lt;antml:function_calls&gt;
&lt;antml:invoke name="format_hello"&gt;
&lt;antml:parameter name="obj">{"type": "individual", "timeOfDay": "morning", "person": "friend"}&lt;/antml:parameter&gt;
&lt;/antml:invoke&gt;
&lt;/antml:function_calls&gt;
```

For the goodbye attempt:
```
&lt;antml:function_calls&gt;
&lt;antml:invoke name="format_goodbye"&gt;
&lt;antml:parameter name="obj">{"type": "individual", "person": "friend", "timeOfDay": "evening"}&lt;/antml:parameter&gt;
&lt;/antml:invoke&gt;
&lt;/antml:function_calls&gt;
```

For the sorry attempt (which worked):
```
&lt;antml:function_calls&gt;
&lt;antml:invoke name="format_sorry"&gt;
&lt;antml:parameter name="obj">{"nested": {"type": "individual", "person": "friend", "timeOfDay": "morning"}}&lt;/antml:parameter&gt;
&lt;/antml:invoke&gt;
&lt;/antml:function_calls&gt;
```

Looking at these now, I can see that the JSON needs to be passed as an actual object, not a string. Let me try the hello format again with the correct approach:

I'm still getting the same error. Would you be able to show me the correct format for these function calls?

---

I'll check the function descriptions in the available tools. Here they are:

For `format_hello`:
```json
{
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "properties": {
      "obj": {
        "anyOf": [
          {
            "additionalProperties": false,
            "properties": {
              "people": {
                "items": {"type": "string"},
                "type": "array"
              },
              "timeOfDay": {
                "description": "The rough time of day, e.g. morning or afternoon",
                "type": "string"
              },
              "type": {"const": "group", "type": "string"}
            },
            "required": ["timeOfDay", "type", "people"],
            "type": "object"
          },
          {
            "additionalProperties": false,
            "properties": {
              "person": {"type": "string"},
              "timeOfDay": {
                "description": "The rough time of day, e.g. morning or afternoon",
                "type": "string"
              },
              "type": {"const": "individual", "type": "string"}
            },
            "required": ["timeOfDay", "type", "person"],
            "type": "object"
          }
        ]
      }
    },
    "required": ["obj"],
    "type": "object"
  }
}
```

For `format_goodbye`:
```json
{
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "properties": {
      "obj": {
        "allOf": [
          {
            "properties": {
              "timeOfDay": {
                "description": "The rough time of day, e.g. morning or afternoon",
                "type": "string"
              }
            },
            "required": ["timeOfDay"],
            "type": "object"
          },
          {
            "anyOf": [
              {
                "additionalProperties": false,
                "properties": {
                  "people": {
                    "items": {"type": "string"},
                    "type": "array"
                  },
                  "type": {"const": "group", "type": "string"}
                },
                "required": ["type", "people"],
                "type": "object"
              },
              {
                "additionalProperties": false,
                "properties": {
                  "person": {"type": "string"},
                  "type": {"const": "individual", "type": "string"}
                },
                "required": ["type", "person"],
                "type": "object"
              }
            ]
          }
        ]
      }
    },
    "required": ["obj"],
    "type": "object"
  }
}
```

For `format_sorry`:
```json
{
  "parameters": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "additionalProperties": false,
    "properties": {
      "obj": {
        "additionalProperties": false,
        "properties": {
          "nested": {
            "anyOf": [
              {
                "additionalProperties": false,
                "properties": {
                  "people": {
                    "items": {"type": "string"},
                    "type": "array"
                  },
                  "timeOfDay": {
                    "description": "The rough time of day, e.g. morning or afternoon",
                    "type": "string"
                  },
                  "type": {"const": "group", "type": "string"}
                },
                "required": ["timeOfDay", "type", "people"],
                "type": "object"
              },
              {
                "additionalProperties": false,
                "properties": {
                  "person": {"type": "string"},
                  "timeOfDay": {
                    "description": "The rough time of day, e.g. morning or afternoon",
                    "type": "string"
                  },
                  "type": {"const": "individual", "type": "string"}
                },
                "required": ["timeOfDay", "type", "person"],
                "type": "object"
              }
            ]
          }
        },
        "required": ["nested"],
        "type": "object"
      }
    },
    "required": ["obj"],
    "type": "object"
  }
}
```

Now I understand better - let me try again with the correct formats:

For hello:

I'm still getting an error despite using the correct schema structure. It seems the parameter needs to be passed differently. Could you show me an example of a correct function call format? The error suggests it's expecting an object but receiving a string, even though I'm providing a JSON object.

**Comment by jspahrsummers** [comment]
I think this is unrelated to MCP, but thank you for the detailed description of the issue and examples. Will pass it along to other teams at Anthropic. Thanks!

**Reply by domdomegg** [reply]
Sounds good, thanks 🙌
